### PR TITLE
Mark `str::is_char_boundary` and `str::split_at*` unstably `const`.

### PR DESCRIPTION
**Pull request by zachs18**
Tracking issues: #131516, #131518

First commit implements `const_is_char_boundary`, second commit implements `const_str_split_at` (which depends on `const_is_char_boundary`)

~~I used `const_eval_select` for `is_char_boundary` since there is a comment about optimizations that would theoretically not happen with the simple `const`-compatible version (since `slice::get` is not `const`ifiable) cc #84751. I have not checked if this code difference is still required for the optimization, so it might not be worth the code complication, but 🤷.~~

This changes `str::split_at_checked` to use a new private helper function `split_at_unchecked` (copied from `split_at_mut_unchecked`) that does pointer stuff instead of `get_unchecked`, since that is not currently `const`ifiable due to using the `SliceIndex` trait.

### CODE REVIEW

**Comment by okaneco** [comment]
I found the commit https://github.com/rust-lang/rust/commit/f621193e5ea7ac54fcd37f0e730e955fd9f61200 that added the optimization comments mentioning avoiding a panic branch and touching memory when index was 0 which you can see here, comparing 1.8, 1.9, and nightly
https://rust.godbolt.org/z/GTx4Mc3G4

I don't think the last comment line rules out const-incompatibility. It looks like it's trying to reinforce that `slice.get(..i)` relies on that specific early return to optimize out checks, not that using `get(i)` in `is_char_boundary` is required for the optimization.

I would assume you'd get the same optimizations if the `match`+`get()` was converted to an `if-else` statement, but there should probably be codegen tests to make sure the 0-constant slicing behavior is unchanged.
```rust
if index == 0 { return true; }
if index < self.as_bytes().len() {
    self.as_bytes()[index].is_utf8_char_boundary()
} else {
    index == self.len()
}
```

**Comment by Noratrieb** [comment]
Yeah, I think you should do the check as @okaneco described, which is const-compatible and should produce the same code in the end.

**Comment by zachs18** [comment]
Latest push translates the prior code as literally as possible to a const-compatible version, without using `const_eval_select`

**Comment by Noratrieb** [comment]
thanks, that looks a lot better.
borsrplus in the style of the ToysRUs logo

File: library/core/src/str/mod.rs [hunk 1]
```diff
@@ -758,7 +762,9 @@ impl str {
     #[inline]
     #[must_use]
     #[stable(feature = "split_at_checked", since = "1.80.0")]
-    pub fn split_at_mut_checked(&mut self, mid: usize) -> Option<(&mut str, &mut str)> {
+    #[rustc_const_unstable(feature = "const_str_split_at", issue = "131518")]
+    #[rustc_allow_const_fn_unstable(const_is_char_boundary)]
```
Comment (RalfJung) [inline library/core/src/str/mod.rs:766]: Why did you add `rustc_allow_const_fn_unstable` here? That makes no sense on an unstable const fn.

Also @Noratrieb please make sure to get wg-const-eval approval before adding any rustc_allow_const_fn_unstable.

Comment (RalfJung) [reply library/core/src/str/mod.rs:766]: Ideally we'd have a bot helping with this attribute policy... https://github.com/rust-lang/triagebot/issues/1851

Comment (Noratrieb) [reply library/core/src/str/mod.rs:766]: ah, I didn't see this, yeah. I was a bit unsure whether I need to ping const eval here but I deemed that there wasn't anything weird enough going on to be problematic (which I guess is sort of true here, the attribute is just misplaced). will be more careful next time

Comment (zachs18) [reply library/core/src/str/mod.rs:766]: > Why did you add `rustc_allow_const_fn_unstable` here? That makes no sense on an unstable const fn. (I'm cleaning this up in #132451.)

I added it because I got an error oyherwise, though that was on an earlier version of the PR.
